apps/web/src/studio/LocalStudioAuthenticator.tsx

```diff
@@ -0,0 +1,67 @@
+import { useEffect } from 'react';
+import { getToken, useAuth } from '../hooks/useAuth';
+import { ROUTES } from '../constants/routes';
+import { encodeBase64 } from './utils/base64';
+import { ILocalStudioState } from './studio.types';
+import { useLocation } from 'react-router-dom';
+
+const BRIDGE_ENDPOINT_PATH = '/api/novu';
+
+function buildStudioURL(state: ILocalStudioState) {
+  const url = new URL(ROUTES.LOCAL_STUDIO, window.location.origin);
+  url.searchParams.append('state', encodeBase64(state));
+
+  return url.href;
+}
+
+export function LocalStudioAuthenticator() {
+  const { currentUser } = useAuth();
+  const location = useLocation();
+
+  useEffect(() => {
+    if (!currentUser) {
+      return;
+    }
+
+    const parsedSearchParams = new URLSearchParams(location.search);
+
+    // Get the redirect URL of the Local Studio server
+    const redirectURL = parsedSearchParams.get('redirect_url');
+
+    if (!redirectURL) {
+      throw new Error('Failed to load Local Studio: missing_redirect_url_parameter.');
+    }
+
+    // Get the optional tunnel origin parameter
+    const tunnelOrigin = parsedSearchParams.get('tunnel_origin');
+
+    const parsedRedirectURL = new URL(redirectURL);
+
+    // Build the state that will be passed to the Local Studio iframe
+    const state = {
+      testUser: {
+        id: currentUser._id,
+        firstName: currentUser.firstName || '',
+        lastName: currentUser.lastName || '',
+        emailAddress: currentUser.email || '',
+        profilePicture: currentUser.profilePicture || '',
+      },
+      authToken: getToken(),
+      localBridgeURL: `${parsedRedirectURL.origin}${BRIDGE_ENDPOINT_PATH}`,
+      tunnelBridgeURL: `${tunnelOrigin}${BRIDGE_ENDPOINT_PATH}`,
+    };
+
+    /*
+     * Construct the final redirect URL poining to the Local Studio server and add
+     * the iframe src URL as a search param.
+     */
+    const finalRedirectURL = new URL(redirectURL);
+    finalRedirectURL.searchParams.append('local_studio_url', buildStudioURL(state));
+
+    // Redirect to Local Studio server
+    window.location.href = finalRedirectURL.href;
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
❓ **question:** Is there a reason we're using this over `navigate` from react-router?‏